src/libsyntax/util/parser_testing.rs

```diff
@@ -83,69 +83,67 @@ pub fn strs_to_idents(ids: Vec<&str> ) -> Vec<ast::Ident> {
 
 /// Does the given string match the pattern? whitespace in the first string
 /// may be deleted or replaced with other whitespace to match the pattern.
-/// this function is Unicode-ignorant; fortunately, the careful design of
-/// UTF-8 mitigates this ignorance.  In particular, this function only collapses
-/// sequences of \n, \r, ' ', and \t, but it should otherwise tolerate Unicode
-/// chars. Unsurprisingly, it doesn't do NKF-normalization(?).
+/// This function is relatively Unicode-ignorant; fortunately, the careful design
+/// of UTF-8 mitigates this ignorance. It doesn't do NKF-normalization(?).
 pub fn matches_codepattern(a : &str, b : &str) -> bool {
-    let mut idx_a = 0;
-    let mut idx_b = 0;
+    let mut a_iter = a.chars().peekable();
+    let mut b_iter = b.chars().peekable();
+
     loop {
-        if idx_a == a.len() && idx_b == b.len() {
-            return true;
-        }
-        else if idx_a == a.len() {return false;}
-        else if idx_b == b.len() {
-            // maybe the stuff left in a is all ws?
-            if is_whitespace(char_at(a, idx_a)) {
-                return scan_for_non_ws_or_end(a,idx_a) == a.len();
-            } else {
-                return false;
+        let (a, b) = match (a_iter.peek(), b_iter.peek()) {
+            (None, None) => return true,
+            (None, _) => return false,
+            (Some(&a), None) => {
+                if is_pattern_whitespace(a) {
+                    break // trailing whitespace check is out of loop for borrowck
+                } else {
+                    return false
+                }
             }
-        }
-        // ws in both given and pattern:
-        else if is_whitespace(char_at(a, idx_a))
-           && is_whitespace(char_at(b, idx_b)) {
-            idx_a = scan_for_non_ws_or_end(a,idx_a);
-            idx_b = scan_for_non_ws_or_end(b,idx_b);
-        }
-        // ws in given only:
-        else if is_whitespace(char_at(a, idx_a)) {
-            idx_a = scan_for_non_ws_or_end(a,idx_a);
-        }
-        // *don't* silently eat ws in expected only.
-        else if char_at(a, idx_a) == char_at(b, idx_b) {
-            idx_a += 1;
-            idx_b += 1;
-        }
-        else {
-            return false;
+            (Some(&a), Some(&b)) => (a, b)
+        };
+
+        if is_pattern_whitespace(a) && is_pattern_whitespace(b) {
+            // skip whitespace for a and b
+            scan_for_non_ws_or_end(&mut a_iter);
+            scan_for_non_ws_or_end(&mut b_iter);
+        } else if is_pattern_whitespace(a) {
+            // skip whitespace for a
+            scan_for_non_ws_or_end(&mut a_iter);
+        } else if a == b {
+            a_iter.next();
+            b_iter.next();
+        } else {
+            return false
         }
     }
+
+    // check if a has *only* trailing whitespace
+    a_iter.all(|c| is_pattern_whitespace(c))
```

Comment:
Is the closure needed? Wouldn't `a_iter.all(is_pattern_whitespace)` work just fine?